Quorum strat update

bas-vk opened the pull request:
This PR includes:

- a bugfix for a race condition that could lead to a corrupt pending state
- change in the block maker/voting strategy that prevents deadlocks
- seed (pseudo) random deadline timers in voting strategy
- output the vote block height instead of the current block number (fixes #26)
- updates the runtime code for the vote contract to the latest solidity version (4.10) @jpmsam, this must also be changed in the quorum example genesis file).
- improves input validation for `getCanonHash`, fixes #44

bas-vk committed:
reimplement block maker and voting strategy

- prevent deadlocks in block maker/vote strategy
- deprecate singleblockmaker flag
- seed pseudoe random block maker/vote deadline strategy
- prevent tx propagation to tx pool during synchronisation

bas-vk committed:
output voted block number and upgrade vote contract to latest solc version

bas-vk committed:
improve input checking for getCanonHash in voting contract

bas-vk commented on eth/handler.go:
```diff
@@ -112,7 +112,7 @@ func NewProtocolManager(config *core.ChainConfig, singleMiner bool, networkId in
 	}
-	if singleMiner {
+	if assumeSyncedInitially {
 		manager.synced = uint32(1)
```
That should not be not be an issue for quorum. In quorum each blockmaker starts with the [assumption](https://github.com/jpmorganchase/quorum/pull/92/files#diff-5b7bdb45b66a5017bf89b46e6ef7a7f4R686) that it is already synced with the chain. If a node finds a node with a better chain it temporary stops accepting transactions until it has caught up.

This isn't an optimal solution and has corner cases that *might* explain why your transactions aren't propagated. Especially in small networks or networks with bad connectivity. It would help to determine what the problem is if you describe what your network setup looks like and what you are seeing.

jpmsam replied:
@hermanjunge have you experienced the reported issue with this PR? I have multiple tests sending transactions across multiple nodes with different roles and I didn't see a TX propagation issue.

bas-vk committed:
inform peers of block head after sync

bas-vk committed:
format voting contract

bas-vk committed:
separate block creation/voting pause/resume

bas-vk committed:
update voting contract runtime code with verifiable bytecode

jpmsam commented on core/quorum/block_voting.go:
```diff
@@ -223,42 +237,67 @@ func (bv *BlockVoting) run(strat BlockMakerStrategy) {
 				case downloader.StartEvent: // begin synchronising, stop block creation and/or voting
-					bv.synced = false
-					strat.Pause()
+					strat.PauseBlockMaking()
```
Please pause voting here too since it has been separated from block voting.

jpmsam commented on core/quorum/block_voting.sol:
```diff
@@ -13,27 +13,27 @@ contract BlockVoting {
     // Raised when a new address is allowed to vote.
```
In the vote event, please change  the name of the param from "blockNumber" to "height" to match the name of the expected value.

jpmsam commented on core/quorum/vote_strategy.go:
```diff
@@ -1,99 +1,156 @@
+		maxBlockTime:      int(maxBlockTime),
+		minVoteTime:       int(minVoteTime),
+		maxVoteTime:       int(maxVoteTime),
+		blockCreateActive: true,
```
The `blockCreateActive` and `voteActive` flags are always set to `true`, can the default values be based on the role of the node? e.g: if the node is configured for voting, then `voteActive` will be `true`, otherwise `false`. I think the voting and block creation timers should only be activated based on the role of the node.

jpmsam reviewed (changes requested):
@bas-vk please review my feedback and update as needed.

bas-vk committed:
various improvements

- stop voting when chain synchronizes
- rename parameter in Vote vent
- immediatly vote when new block is imported
- activate vote/block creation when node is configured for a particular role

jpmsam commented on core/quorum/vote_strategy.go:
```diff
@@ -1,99 +1,168 @@
 	}
-	s.deadlineTimer = time.NewTimer(time.Duration(s.min+rand.Intn(s.max-s.min)) * time.Second)
+
+	s.voteTimer = time.NewTimer(time.Duration(s.minBlockTime+rand.Intn(s.maxVoteTime-s.minVoteTime)) * time.Second)
```
@bas-vk based on my earlier suggestion, can you please create the timer for voting and block making only if the node is configured with votingActive or blockCreateActive? I don't see a reason why the timer should be active if the node can't vote or create blocks.

jpmsam reviewed (changes requested):
Hi @bas-vk, thanks for the latest commit with the fixes. I am testing the strategy update but please see my note regarding the timers.
